features/nanostack/mbed-mesh-api/source/NanostackPPPInterface.cpp

```diff
@@ -0,0 +1,402 @@
+    } else if (status != 0) {
+        return NSAPI_ERROR_DEVICE_ERROR;
+    }
+
+    if (_blocking) {
+        int32_t count = disconnect_semaphore.try_acquire_for(30000);
+
+        if (count <= 0) {
+            return NSAPI_ERROR_TIMEOUT;
+        }
+    }
+    return NSAPI_ERROR_OK;
+}
+
+void Nanostack::PPPInterface::link_state_changed(bool up)
+{
+    // Read negotiated parameters from PPP
+    if (up && !link_state_up) {
+        PPPPhy &phy = static_cast<PPPPhy &>(get_phy());
+        phy.get_negotiated_parameters();
+    }
+
+    link_state_up = up;
+
+    // If application callback is set
+    if (link_state_cb) {
+        link_state_cb(up, _device_id);
+    } else {
+        if (up) {
+            connect_enet_tasklet();
+        }
+    }
+}
+
+void Nanostack::PPPInterface::set_link_state_changed_callback(link_state_cb_t new_link_state_cb)
+{
+    link_state_cb = new_link_state_cb;
+}
+
+// GAH! no handles on the callback. Force a single interface
+static PPPPhy *single_phy;
+
+extern "C"
+{
+    static int8_t ppp_phy_address_write(phy_address_type_e address_type, uint8_t *address_ptr)
+    {
+        return single_phy->address_write(address_type, address_ptr);
+    }
+
+    static int8_t ppp_phy_interface_state_control(phy_interface_state_e, uint8_t)
+    {
+        return -1;
+    }
+
+    static int8_t ppp_phy_tx(uint8_t *data_ptr, uint16_t data_len, uint8_t tx_handle, data_protocol_e data_flow)
+    {
+        return single_phy->tx(data_ptr, data_len, tx_handle, data_flow);
+    }
+
+    PPPPhy::PPPPhy(NanostackMemoryManager &mem, PPP &m) : memory_manager(mem), ppp(m), link_state_change_cb(NULL), active(false), powered_up(false), device_id(-1)
+    {
+    }
+
+    void PPPPhy::ppp_phy_rx(net_stack_mem_buf_t *mem)
+    {
+        if (!active) {
+            memory_manager.free(mem);
+            return;
+        }
+
+        const uint8_t *ptr = NULL;
+        uint8_t *tmpbuf = NULL;
+        uint32_t total_len;
+
+        if (memory_manager.get_next(mem) == NULL) {
+            // Easy contiguous case
+            ptr = static_cast<const uint8_t *>(memory_manager.get_ptr(mem));
+            total_len = memory_manager.get_len(mem);
+        } else {
+            // Nanostack can't accept chunked data - make temporary contiguous copy
+            total_len = memory_manager.get_total_len(mem);
+            ptr = tmpbuf = static_cast<uint8_t *>(ns_dyn_mem_temporary_alloc(total_len));
+            if (tmpbuf) {
+                memory_manager.copy_from_buf(tmpbuf, total_len, mem);
+            }
+        }
+
+        if (ptr && phy.phy_rx_cb) {
+            phy.phy_rx_cb(ptr, total_len, 0xff, 0, device_id);
+        }
+        ns_dyn_mem_free(tmpbuf);
+        memory_manager.free(mem);
+    }
+
+} // extern "C"
+
+int8_t PPPPhy::address_write(phy_address_type_e address_type, uint8_t *address_ptr)
+{
+    if (address_type != PHY_MAC_64BIT) {
+        return -1;
+    }
+    //memcpy(unique_id, address_ptr, 8);
```

Comment:
Are these commented codes needed?

Reply:
Removed the write function since not valid for ppp.